embassy-stm32/src/dac/v2.rs

```diff
@@ -0,0 +1,295 @@
+use crate::dac::{DacPin, Instance};
+use crate::gpio::Pin;
+use crate::gpio::{AnyPin, OptionalPin};
+use crate::pac::regs::dac_v2;
+use core::marker::PhantomData;
+use embassy::util::Unborrow;
+use embassy_extras::unborrow;
+
+pub enum Error {
+    UnconfiguredChannel,
+    InvalidValue,
+}
+
+pub enum Ch1Trigger {
+    Tim6,
+    Tim3,
+    Tim7,
+    Tim15,
+    Tim2,
+    Exti9,
+    Software,
+}
+
+impl Ch1Trigger {
+    fn tsel(&self) -> dac_v2::vals::Tsel1 {
+        match self {
+            Ch1Trigger::Tim6 => dac_v2::vals::Tsel1::TIM6_TRGO,
+            Ch1Trigger::Tim3 => dac_v2::vals::Tsel1::TIM3_TRGO,
+            Ch1Trigger::Tim7 => dac_v2::vals::Tsel1::TIM7_TRGO,
+            Ch1Trigger::Tim15 => dac_v2::vals::Tsel1::TIM15_TRGO,
+            Ch1Trigger::Tim2 => dac_v2::vals::Tsel1::TIM2_TRGO,
+            Ch1Trigger::Exti9 => dac_v2::vals::Tsel1::EXTI9,
+            Ch1Trigger::Software => dac_v2::vals::Tsel1::SOFTWARE,
+        }
+    }
+}
+
+pub enum Ch2Trigger {
+    Tim6,
+    Tim8,
+    Tim7,
+    Tim5,
+    Tim2,
+    Tim4,
+    Exti9,
+    Software,
+}
+
+impl Ch2Trigger {
+    fn tsel(&self) -> dac_v2::vals::Tsel2 {
+        match self {
+            Ch2Trigger::Tim6 => dac_v2::vals::Tsel2::TIM6_TRGO,
+            Ch2Trigger::Tim8 => dac_v2::vals::Tsel2::TIM8_TRGO,
+            Ch2Trigger::Tim7 => dac_v2::vals::Tsel2::TIM7_TRGO,
+            Ch2Trigger::Tim5 => dac_v2::vals::Tsel2::TIM5_TRGO,
+            Ch2Trigger::Tim2 => dac_v2::vals::Tsel2::TIM2_TRGO,
+            Ch2Trigger::Tim4 => dac_v2::vals::Tsel2::TIM4_TRGO,
+            Ch2Trigger::Exti9 => dac_v2::vals::Tsel2::EXTI9,
+            Ch2Trigger::Software => dac_v2::vals::Tsel2::SOFTWARE,
+        }
+    }
+}
+
+pub enum Alignment {
+    Left,
+    Right,
+}
+
+pub enum Value {
+    Bit8(u8),
+    Bit12(u16, Alignment),
+}
+
+pub struct Dac<'d, T: Instance> {
+    //peri: T,
+    ch1: Option<AnyPin>,
+    ch2: Option<AnyPin>,
+    phantom: PhantomData<&'d mut T>,
+}
+
+impl<'d, T: Instance> Dac<'d, T> {
+    pub fn new(
+        peri: impl Unborrow<Target = T> + 'd,
+        ch1: impl Unborrow<Target = impl DacPin<T, 1>>,
+        ch2: impl Unborrow<Target = impl DacPin<T, 2>>,
```

Comment:
To avoid 3 functions it'd be nicer to use OptionalPin.

- `trait DacPin<..>: OptionalPin`
- impl DacPin for NoPin
- You can then do `.degrade_optional()` to get Option<AnyPin> and go from there.